### PR TITLE
Fixes sign tool for looking nested file extension in SignableExtension list

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: #977 

**Changes:**

- Configuration.cs : look into "SignableExtensions" for nested files as well.
- SignToolTests.cs: path to new input test case.
- SignToolTask.cs: remove requirement for always specifying `StrongNameSignInfo`. Add some validation.

### CODE REVIEW

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 1]
```diff
@@ -333,9 +333,10 @@ private bool TryBuildZipData(FileSignInfo zipFileSignInfo, out ZipData zipData)
                         string relativePath = entry.FullName;
                         string extension = Path.GetExtension(relativePath);
 
-                        if (!FileSignInfo.IsZipContainer(relativePath) && (!_fileExtensionSignInfo.TryGetValue(extension, out var extensionSignInfo) || !extensionSignInfo.ShouldSign))
+                        if (!FileSignInfo.IsZipContainer(relativePath) &&
+                            (!SignToolConstants.SignableExtensions.Contains(extension) || (_fileExtensionSignInfo.TryGetValue(extension, out var extensionSignInfo) && !extensionSignInfo.ShouldSign)))
```
Comment (tmat) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs]: >(_fileExtensionSignInfo.TryGetValue(extension, out var extensionSignInfo) && !extensionSignInfo.ShouldSign [](start = 90, length = 106)

I think we can remove the entire `if` statement. We decide what the sign info for a file is in `TrackFile`. Don't need to do it as well.

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 2]
```diff
@@ -347,11 +347,11 @@ private bool TryBuildZipData(FileSignInfo zipFileSignInfo, out ZipData zipData)
                         }
 
                         // if we already encountered file that hash the same content we can reuse its signed version when repackaging the container.
-                        string fileName = Path.GetFileName(relativePath);
+                        string fileName = Path.GetFileName(entry.FullName);
```
Comment (tmat) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs]: >(entry.FullName [](start = 58, length = 15)

why removing the relativePath local? it makes it more clear imo

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 3]
```diff
@@ -330,13 +334,9 @@ private bool TryBuildZipData(FileSignInfo zipFileSignInfo, out ZipData zipData)
 
                     foreach (ZipArchiveEntry entry in archive.Entries)
                     {
-                        string relativePath = entry.FullName;
-                        string extension = Path.GetExtension(relativePath);
-
-                        if (!FileSignInfo.IsZipContainer(relativePath) && (!_fileExtensionSignInfo.TryGetValue(extension, out var extensionSignInfo) || !extensionSignInfo.ShouldSign))
+                        // `entry` might be just a pointer to a folder. We skip those.
+                        if (string.IsNullOrEmpty(Path.GetExtension(entry.FullName)))
```
Comment (tmat) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs]: >Null [](start = 37, length = 4)

GetExtension doesn't return null, so there is no need to check for null

Comment (tmat) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: Actually, this is not quite correct - there might be files with no extension. We should check for folder in another way.

---
In reply to: [224873691](https://github.com/dotnet/arcade/pull/1011#discussion_r224873691) [](ancestors = 224873691)

Comment (tmat) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: [https://stackoverflow.com/questions/40223451/how-to-tell-if-a-ziparchiveentry-is-directory](https://stackoverflow.com/questions/40223451/how-to-tell-if-a-ziparchiveentry-is-directory)

---
In reply to: [224874292](https://github.com/dotnet/arcade/pull/1011#discussion_r224874292) [](ancestors = 224874292,224873691)

Comment (tmat) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: Why would they not make sense? They should be processed like any other file. They could potentially be signed if `FileSignInfo` specifies a certificate for the file name.

Comment (JohnTortugo) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: Oks, sounds good to me.

Comment (JohnTortugo) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: Just pushed new changes.

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs (hunk 2 above)
Comment (tmat) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: Why do you keep changing it? There are still multiple uses of `entry.FullName` just keep the `relativePath` variable as it was.

---
In reply to: [224876776](https://github.com/dotnet/arcade/pull/1011#discussion_r224876776) [](ancestors = 224876776)

Comment (JohnTortugo) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: I didn't change it again... just hadn't put it back yet. Just pushed new changes.

**Comment by tmat** [comment]
Thanks!

**Pull request merged**